compiler/src/dotty/tools/dotc/transform/IsInstanceOfEvaluator.scala

```diff
@@ -147,7 +185,7 @@ class IsInstanceOfEvaluator extends MiniPhaseTransform { thisTransformer =>
 
           val inMatch = s.qualifier.symbol is Case
```

Comment:
@odersky: This logic was proposed by @DarkDimius when I implemented the original `IsInstanceOfEvaluator`, he could perhaps know of some reason why this isn't a problem. But the basic idea was that:

`val` bindings created inside a pattern match desugaring would get the `case` flag. This was so that we can know that they were specifically created by the `PatternMatcher`. 

I'd have to verify that this does indeed conflict with `case object`s. 

If you're sure (or if I get to verifying before you respond), then if I've understood your proposal correctly:

- `PatternMatcher` should be changed to emit a special `caseIsInstanceOf` symbol (or similar)
- `IsInstanceOfEvaluator` should intercept these and replace them with either regular `isInstanceOf` checks or the evaluated expressions.

    should it **exclusively** do this for the special symbol, or do we still want the optimization to be performed elsewhere?


Last - do you want this fix included in this PR or can it be separate? The PR doesn't otherwise touch the original logic. I'd open a second PR  with the implementation during the day so that we don't forget.